contracts/peripheral/YieldDirectorV2.sol

```diff
@@ -0,0 +1,451 @@
+    }
+
+    /**
+        @notice Return total amount of gOHM donated to recipient since last full withdrawal or redemption
+        @param donor_ Address of user donating yield
+        @param recipient_ Address of user recieiving donated yield
+     */
+    function donatedTo(address donor_, address recipient_) external view returns (uint256) {
+        uint256[] storage depositIds = depositorIds[donor_];
+
+        for (uint256 index = 0; index < depositIds.length; index++) {
+            if (recipientLookup[depositIds[index]] == recipient_) {
+                return redeemableBalance(depositIds[index]);
+            }
+        }
+
+        return 0;
+    }
+
+    /**
+        @notice Return total amount of gOHM donated from donor since last full withdrawal or redemption
+        @param donor_ Address of user donating yield
+     */
+    function totalDonated(address donor_) external view returns (uint256) {
+        uint256[] storage depositIds = depositorIds[donor_];
+        uint256 principalTotal = 0;
+        uint256 agnosticTotal = 0;
+
+        for (uint256 index = 0; index < depositIds.length; index++) {
+            DepositInfo storage currDeposit = depositInfo[depositIds[index]];
+
+            principalTotal += currDeposit.principalAmount;
+            agnosticTotal += currDeposit.agnosticAmount;
+        }
+
+        return _getOutstandingYield(principalTotal, agnosticTotal);
+    }
+
+    /************************
+     * Recipient Functions
+     ************************/
+
+    function redeemableBalance(uint256 id_) public view returns (uint256) {
+        DepositInfo storage currDeposit = depositInfo[id_];
+        return _getOutstandingYield(currDeposit.principalAmount, currDeposit.agnosticAmount);
+    }
+
+    /**
+        @notice Get redeemable sOHM balance of a recipient address
+        @param recipient_ Address of user receiving donated yield
+     */
+    function totalRedeemableBalance(address recipient_) public view returns (uint256) {
+        uint256[] storage receiptIds = recipientIds[recipient_];
+        uint256 agnosticRedeemable = 0;
+
+        for (uint256 index = 0; index < receiptIds.length; index++) {
+            DepositInfo storage currDeposit = depositInfo[receiptIds[index]];
+            agnosticRedeemable += _getOutstandingYield(currDeposit.principalAmount, currDeposit.agnosticAmount);
+        }
+
+        return agnosticRedeemable;
+    }
+
+    /**
+        @notice Redeem recipient's donated amount of sOHM at current index from one donor as gOHM
+        @param id_ Deposit id for this donation
+    */
+    function redeemYield(uint256 id_) external returns (uint256) {
+        if (redeemDisabled) revert YieldDirector_RedeemsDisabled();
+        require(recipientLookup[id_] == msg.sender);
+
+        uint256 amountRedeemed = _redeemYield(id_);
+        if (amountRedeemed == 0) revert YieldDirector_NoRedeemableBalance();
+
+        IERC20(gOHM).safeTransfer(msg.sender, amountRedeemed);
+
```

Comment:
Is here where you want to add the logic:
if principal == 0. remove deposit from recipient array and deposit to recipient lookup ?

Reply:
Also can we move this piece of code:
if (userDeposit.principalAmount == 0) {
            _closeDeposit(id_);
        }
 From base contract to implementation. I think it might cause some issues for some implementations. I think since implementation handles the recipient logic they should probably also handle closing deposits when they need to.

Reply:
yes think so, and good call. I'll do that.